chore: add editorconfig

SethFalco opened the pull request:
While developing for this, there can be some annoyances with different code styles.
Namely, tabs vs spaces. ^-^'

Could an EditorConfig file be included? It has fairly wide support, so when contributors want to do their thing, the editor settings will automatically apply the conventions already used in the repository.

The `_config.yml` file is just so GitHub Pages doesn't host it on the site.

SethFalco committed:
chore: add editorconfig

LeaVerou commented on .editorconfig:
```diff
@@ -0,0 +1,14 @@
+
+[*]
+indent_style = tab
+indent_size = 2
```
Not sure why we should specify this, the whole point of using tabs is that everyone can use their own tab width. Tab width is a local setting, that does not affect the underlying code at all.

SethFalco replied:
Happy to remove it from in a bit.

I think a common reason to include it is for when max line length is set, it keeps things consistent. For example, if the repo mandates a max line length of 120, then the "length" of lines can vary depending on the indent size of developers code editors.

Ofc that doesn't apply here, honestly… I just left by mistake when copy and pasting it over from another project.

LeaVerou replied:
Even if a line length was set, I think the flexibility of everyone having their preferred tab width outweighs the importance of having shorter line lengths (I could even argue that indentation should not count in line length, though having a small cyclomatic complexity is desirable, and that correlates with fewer indentation levels).

LeaVerou commented on .editorconfig:
```diff
@@ -0,0 +1,14 @@
+root = true
+
+[*]
+indent_style = tab
```
We are actually using smart tabs, not just tabs, but not sure what's the current best practice to specify this in `.editorconfig`. Relevant: https://github.com/editorconfig/editorconfig/issues/323

SethFalco replied:
Hmm, I'm not familiar with smart tabs. 🤔 
I'll look into it and see how that works with EditorConfig, sorry about that.

LeaVerou replied:
Well, some people say that smart tabs is just how tabs *should* be used anyway (as you will see in that discussion as well). Basically, tabs for indentation and spaces for alignment. That way everyone gets to have their preferred tab width, and things that should be aligned, stay aligned.

LeaVerou commented on .editorconfig:
```diff
@@ -0,0 +1,14 @@
+spaces_around_operators = true
+
+[*.{yaml,yml}]
+indent_style = space
```
Why?

SethFalco replied:
YAML mandates spaces!

>  Tabs have been outlawed since they are treated differently by different editors and tools. And since indentation is so critical to proper interpretation of YAML, this issue is just too tricky to even attempt. Indeed Guido van Rossum of Python has acknowledged that allowing TABs in Python source is a headache for many people and that were he to design Python again, he would forbid them.
> 
> — https://yaml.org/faq.html

LeaVerou replied:
Oh ok!

LeaVerou reviewed (changes requested):
Hey, thanks! I left some comments.

Also, I had no idea you could exclude things from GitHub Pages! That's amazing, I've needed it so much in a recent project. Do you have any pointers where I can read more about this feature?

SethFalco commented:
> Also, I had no idea you could exclude things from GitHub Pages! That's amazing, I've needed it so much in a recent project. Do you have any pointers where I can read more about this feature?

GitHub Pages ultimately uses Jekyll by default under-the-hood.

> You can configure most Jekyll settings, such as your site's theme and plugins, by editing your _config.yml file.

> By default, Jekyll doesn't build files or folders that:
> 
> * are located in a folder called /node_modules or /vendor
> * start with `_`, `.`, or `#`
> * end with `~`
> * are excluded by the `exclude` setting in your configuration file

https://docs.github.com/en/pages/setting-up-a-github-pages-site-with-jekyll/about-github-pages-and-jekyll

Then over on the Jekyll documentation, we can see the `exclude` property.

https://jekyllrb.com/docs/configuration/options/

LeaVerou commented:
Thank you so much! So, according to that, dotfiles are already excluded, so why do we need to exclude them via `exclude` as well?

SethFalco commented:
> so why do we need to exclude them via

Because I'm an idiot and despite citing that very information just now I didn't realize what I did was entirely redundant. 😆 
In other words, I forgot `.editorconfig` started with a `.`. 🤦🏾

SethFalco commented:
Resolved all the issues except `indent_style`. That one I just removed as I wasn't sure what the best course of action was.

What you described smart tabs to be matches my experience with just tabs, so I didn't realize it differed. Now that I know this… I'm unsure of how this property impacts editors that differentiate "tabs" vs "smart tabs". (i.e. [IntelliJ](https://www.jetbrains.com/help/idea/code-style-sql-tabs-and-indents.html))

I guess it's best to omit it to err on the side of caution. Just means developers who use spaces would have to override the setting in this repo manually. My fault for being a spaces person. 🥲 

I'm hoping maybe we can still commit this as is, and in future if there is some clarity on EditorConfig defining `smart_tabs` that can be added? I can see a vote passed saying it should be added, and some implementations voluntarily support this property. However, it was not added to the spec for some reason. 🤔 

Though I will admit, `indent_style` was the biggest motivation for this one. ^-^'

LeaVerou commented:
What is the downside of leaving `indent_style: tabs`? Does it complain about mixed tabs and spaces if spaces have been used for alignment?

E.g. in the following

```js
if (foo) {
	var a = 1;
	    b = 2;
}
```

SethFalco commented:
From my experience with Visual Studio Code (VSC), it will not complain about it. I haven't checked in other tools, but I'd be surprised if they would complain about this either if found in existing code or typed manually.

However, auto-formatting will breach the rule and require manual fixing, or replace the spaces for alignment with tabs if the formatter is executed manually.

For example, when on this line:

> ```js
> 		var onBlack = this.overlayOn(_.BLACK),
> 		    onWhite = this.overlayOn(_.WHITE),
> 		    contrastOnBlack = onBlack.contrast(color).ratio,
> ```
>
> — https://github.com/LeaVerou/contrast-ratio/blob/gh-pages/color.js#L145

If I were to press `<ENTER>` here, assuming `indent_style: tab` is configured, it will insert tabs to get to the necessary alignment. VSC is happy to allow both tabs and spaces in this mode, but the way it inserts them automatically doesn't match the rule of "tabs for indentation" and "spaces for alignment".

In my case, I have my indentation size set to 2 for JS. So, it sees the 4 spaces above, and then automatically put 2 tabs.

Overall, if I wouldn't manually correct it, the result would be:

```js
		var onBlack = this.overlayOn(_.BLACK),
		    onWhite = this.overlayOn(_.WHITE),
		    contrastOnBlack = onBlack.contrast(color).ratio,
				contrastOnWhite = onWhite.contrast(color).ratio;
```

(In my editor it'll look perfectly fine since I visually have tabs to have a width of 2, but it doesn't translate over well to GitHub for example.)

Edit: I made a 2nd remark which I've deleted about what happens when the spaces for alignment isn't divisible by the visual indentation size of the editor. If you saw it, just ignore that. VSC does handle that correctly and does tabs for indentation and spaces for alignment in those cases. I just forgot I removed the line from the `.editorconfig` so it was set to use spaces instead of tabs.

SethFalco commented:
The above I don't consider a bug, just a bit faff because of the mismatch of style vs the convention in the file. Just thought I'd share another thing I tested, which I would call a bug. This is more for fun rather than meaningful input, though. ^-^'

If I set my indentation size to where the padded spaces for alignment isn't divisible by the indentation size. (i.e. where there are 4 spaces for alignment, and I visually have an indentation size of 6)

VSC opts to just put spaces the entire way through instead of tabs for indentation and spaces for alignment, so there are no tabs on the new line at all setting in the editor that I use tabs. ^-^'

```js
		var onBlack = this.overlayOn(_.BLACK),
		    onWhite = this.overlayOn(_.WHITE),
		    contrastOnBlack = onBlack.contrast(color).ratio,
                contrastOnWhite = onWhite.contrast(color).ratio;
```

LeaVerou commented:
I see. Is that behavior different than when no indent style is configured and the editor just matches what's in the file?

SethFalco commented:
> Is that behavior different than when no indent style is configured and the editor just matches what's in the file?

I can't comment on that specifically. I can say if it's not configured, the code editor will just handle indentation as it would've done if the `.editorconfig` file wasn't there at all. Ultimately, that comes down to what the developer configured in their editor's preferences, though.

In VSC using the default settings, and no installed extensions, it'd probably do the same thing as above. (Auto-detect indentation is set by default, and without extensions there is no smart tab support afaik.)

Meanwhile, in both VSC and Vim I personally have it set to always use spaces unless overridden otherwise. Usually EditorConfig will handle this for me, but if not, then I just do a command to stop expanding tabs. (Though, we can see from this PR that I forgot… My fault, when I moved the `.editorconfig` to a separate branch, I forgot that'd undo the setting and didn't update workspace settings.)

Worrying about multiple editors is actually why I like EditorConfig a bit, keeping the conventions in a single editor-agnostic file seems handy for me. Though ofc, it's less handy if it doesn't support the convention one wants to follow. ^-^'

LeaVerou commented:
Ok, I'll leave it up to you if we should have `indent_style: tabs` or not. Btw I don't think there's any editor that does auto-formatting with smart tabs correctly.

SethFalco commented:
I think it's fine to leave it out for now. Once EditorConfig make up their mind about `smart_tabs`, and it's either in the [spec](https://editorconfig-specification.readthedocs.io/#supported-pairs) or documented in the [wiki](https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties), we can look at this again.

So, long as you're happy with everything else, I think it's fine to merge this.

LeaVerou commented:
Merged, thanks!
